### ISSUE TITLE
mut_from_ref and Pin<&mut>

### ISSUE THREAD

**Issue by geeklint** [rust-lang/rust-clippy#7749]
Lint name:  mut_from_ref

I tried this code:

```rust
use std::pin::Pin;

pub struct IsUnpin {
    data: u32,
    _marker: std::marker::PhantomPinned,
}

impl IsUnpin {
    pub fn test<'a>(self: Pin<&'a mut Self>, _something_else: &'a ()) -> &'a mut u32 {
        unsafe {
            &mut self.get_unchecked_mut().data
        }
    }
}
```

I expected to see this happen: No errors

Instead, this happened: Clippy reported the `mut_from_ref` error level lint.

The lint documentation says

> To be on the conservative side, if there’s at least one mutable reference with the output lifetime, this lint will not trigger.

and the lint does not trigger if it is just `Pin<&mut Self>` by itself, only when it is used in conjunction with another immutable borrow.  If there is a reason this is not a false positive, can someone explain the details to me?

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.55.0 (c8dfcfe04 2021-09-06)
binary: rustc
commit-hash: c8dfcfe046a7680554bf4eb612bad840e7631c4b
commit-date: 2021-09-06
host: x86_64-unknown-linux-gnu
release: 1.55.0
LLVM version: 12.0.1
```

@ rustbot label +I-suggestion-causes-error

**Comment by xFrednet** [comment]
This looks like a false positive. From the example, I would be guessing that the logic only checks if `self` is `mut` and not what the type of self is (I can be wrong though)

**Comment by Xaeroxe** [comment]
Also I just realized this lint isn't recursing into tuples, slices, or arrays.

**Comment by Xaeroxe** [comment]
I found this: https://github.com/rust-lang/rust-clippy/blob/master/clippy_utils/src/ty.rs#L118 which seems useful. I also learned that walking a type returns lifetime information, now I just need to determine a way to make that lifetime information useful in this context.
